[jit] get get/setstate working in new format

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23308 [jit] get get/setstate working in new format**
* #23098 [wip][jit] serialize modules as classes
* #23154 [jit] Put all modules in the global Python CU
* #23170 [jit] Make optimize a thread_local flag

This needs to be squashed with the previous PR and landed together, but
I'm pulling it out for review

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23308

Differential Revision: [D16460632](https://our.internmc.facebook.com/intern/diff/D16460632)

suo committed:
[jit] get get/setstate working in new format

This needs to be squashed with the previous PR and landed together, but
I'm pulling it out for review

suo commented on torch/csrc/jit/script/python_sugared_value.cpp:
```diff
@@ -239,6 +239,10 @@ std::shared_ptr<SugaredValue> OverloadedMethodValue::call(
                          << err.str();
 }
 
+Value* ModuleValue::asValue(const SourceRange& loc, Function& m) {
```
@zdevito do you think this is safe to add now that modules are first-class?

zdevito replied:
Yes, it should be there. I am surprised that stuff worked without it actually.

suo committed:
Update on "[jit] get get/setstate working in new format"


This needs to be squashed with the previous PR and landed together, but
I'm pulling it out for review

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23308

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -691,11 +691,56 @@ struct to_ir {
+      Block* block) {
+    std::vector<Argument> arguments; // for schema
+    TORCH_INTERNAL_ASSERT(def.name().name() == "__setstate__");
+    TORCH_INTERNAL_ASSERT(def.decl().params().size() == 2);
```
This is not an internal  assert?

suo committed:
Update on "[jit] get get/setstate working in new format"


This needs to be squashed with the previous PR and landed together, but
I'm pulling it out for review

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23308

Differential Revision: [D16460632](https://our.internmc.facebook.com/intern/diff/D16460632)

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -691,11 +691,56 @@ struct to_ir {
+      const auto self_param = def.decl().params()[0];
+      const auto& name = self_param.ident().name();
+      Value* new_input = block->addInput()->setDebugName(name);
+      environment_stack->setSugaredVar(
```
I am not a fan of the code duplication here, in particular the environment stack handling. This function should just be providing a type to the argument of  __getstate__ not also duplciating the logic that puts self as a sugared var and initializes the environment. If either of these things change, someone will forget to edit this code.

suo replied:
makes sense

zdevito reviewed (approved):
Looks good, minor nit about code duplication.

suo committed:
Update on "[jit] get get/setstate working in new format"


This needs to be squashed with the previous PR and landed together, but
I'm pulling it out for review

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23308

Differential Revision: [D16460632](https://our.internmc.facebook.com/intern/diff/D16460632)

suo committed:
Update on "[jit] get get/setstate working in new format"


This needs to be squashed with the previous PR and landed together, but
I'm pulling it out for review

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23308

Differential Revision: [D16460632](https://our.internmc.facebook.com/intern/diff/D16460632)

suo commented:
closing as this has been squashed with https://github.com/pytorch/pytorch/pull/23098
